### PR TITLE
feat(expr): add `vnode` internal function

### PR DESCRIPTION
**Pull request by richardchien**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Add `vnode` function.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

#4247

### CODE REVIEW

**Comment by fuyufjh** [comment]
Need to refactor to variable vnode after #3396

**Review by fuyufjh — APPROVED** [review]
LGTM.

By the way, should we expose it as a function for debug purpose? cc. @st1page

**Comment by richardchien** [comment]
> By the way, should we expose it as a function for debug purpose? cc. @st1page 

Exposing it as `_vnode` sounds good to me.

File: src/expr/src/expr/expr_vnode.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            }
+
+            let dist_key = dist_key_columns
+                .iter()
+                .map(|col| col.datum_at(row_idx))
+                .collect();
+            let dist_key_row = Row::new(dist_key);
+            let vnode = dist_key_row.hash_row(&CRC32FastBuilder {}).to_vnode() as i32;
+            builder.append(Some(vnode))?;
+        }
```
Comment (BugenZhao) [inline src/expr/src/expr/expr_vnode.rs:89]: I think we should avoid collecting owned rows here and reuse `StreamChunk::get_hash_values`, which is in a more columnar style. May check hash dispatcher for references.

https://github.com/singularity-data/risingwave/blob/5510952c00bc73922ec71b1dde06274fd6ad179a/src/stream/src/executor/dispatch.rs#L532-L537

File: src/expr/src/expr/expr_vnode.rs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            .map(expr_build_from_prost)
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(VnodeExpression::new(dist_key_exprs))
+    }
+}
+
+impl Expression for VnodeExpression {
+    fn return_type(&self) -> DataType {
+        DataType::Int32
```
Comment (BugenZhao) [inline src/expr/src/expr/expr_vnode.rs:62]: Int16 should be enough?
```suggestion
        DataType::Int16
```

**Review by BugenZhao — CHANGES_REQUESTED** [review]

File: src/expr/src/expr/expr_vnode.rs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+        let RexNode::FuncCall(func_call_node) = prost.get_rex_node().unwrap() else {
+            bail!("Expected RexNode::FuncCall");
+        };
+
+        let dist_key_exprs = func_call_node
+            .children
+            .iter()
+            .map(expr_build_from_prost)
+            .collect::<Result<Vec<_>>>()?;
```
Comment (BugenZhao) [inline src/expr/src/expr/expr_vnode.rs:54]:
```suggestion
            .try_collect()?;
```

File: src/expr/src/expr/expr_vnode.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    fn eval_row(&self, input: &Row) -> Result<Datum> {
+        let dist_key = self
+            .dist_key_exprs
+            .iter()
+            .map(|c| c.eval_row(input))
+            .collect::<Result<Vec<_>>>()?;
+        let dist_key_row = Row::new(dist_key);
+        let vnode = dist_key_row.hash_row(&CRC32FastBuilder {}).to_vnode() as i32;
```
Comment (BugenZhao) [inline src/expr/src/expr/expr_vnode.rs:101]: May add a `FIXME` or note here: currently the implementation of the hash function in `Row::hash_row` differs from `Array::hash_at`, so their result might be different. #3457

File: src/expr/src/expr/expr_vnode.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+use risingwave_common::util::hash_util::CRC32FastBuilder;
+use risingwave_pb::expr::expr_node::{RexNode, Type};
+use risingwave_pb::expr::ExprNode;
+
+use super::{build_from_prost as expr_build_from_prost, BoxedExpression, Expression};
+use crate::{bail, ensure, ExprError, Result};
+
+#[derive(Debug)]
+pub struct VnodeExpression {
+    dist_key_exprs: Vec<BoxedExpression>,
```
Comment (st1page) [inline src/expr/src/expr/expr_vnode.rs:31]: why not use `Vec<usize>`?

**Review by st1page — APPROVED** [review]
LGTM

**Review by BugenZhao — APPROVED** [review]
LGTM

**Comment by richardchien** [comment]
@mergifyio requeue

**Pull request merged**